File: core/src/main/kotlin/br/com/dillmann/nginxignition/core/common/dynamicfield/DynamicField.kt
package br.com.dillmann.nginxignition.core.common.dynamicfield

data class DynamicField(
    val id: String,
    val priority: Int,
    val description: String,
    val required: Boolean,
    val type: Type,
    val enumOptions: List<EnumOption> = emptyList(),
    val sensitive: Boolean = false,
    val condition: Condition? = null,
    val helpText: String? = null,
    val defaultValue: String? = null,
) {
    data class EnumOption(
        val id: String,
        val description: String,
    )

    data class Condition(
        val parentField: String,
        val value: String,
    )

    enum class Type {
        SINGLE_LINE_TEXT,
        MULTI_LINE_TEXT,
        URL,
        EMAIL,
        BOOLEAN,
        ENUM,
        FILE,
    }
}

File: frontend/src/core/dynamicfield/DynamicField.ts
export default interface DynamicField {
    id: string
    priority: number
    description: string
    required: boolean
    sensitive: boolean
    type: DynamicFieldType
    enumOptions: DynamicFieldEnumOption[]
    conditions?: DynamicFieldCondition[]
    helpText?: string
    defaultValue?: string
}

export enum DynamicFieldType {
    SINGLE_LINE_TEXT = "SINGLE_LINE_TEXT",
    MULTI_LINE_TEXT = "MULTI_LINE_TEXT",
    URL = "URL",
    EMAIL = "EMAIL",
    BOOLEAN = "BOOLEAN",
    ENUM = "ENUM",
    FILE = "FILE",
}

export interface DynamicFieldEnumOption {
    id: string
    description: string
}

export interface DynamicFieldCondition {
    parentField: string
    value: any
}
